src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/CpsDiagnosticItemSourceProvider.cs

```diff
@@ -100,66 +101,61 @@ internal sealed class CpsDiagnosticItemSourceProvider : AttachedCollectionSource
         /// </summary>
         private static string? GetTargetFrameworkMoniker(IVsHierarchyItem item)
         {
-            var hierarchy = item.HierarchyIdentity.NestedHierarchy;
-            var itemId = item.HierarchyIdentity.NestedItemID;
-
-            var isTargetNode = false;
-            string? potentialTFM = null;
-
-            MatchProjectTreeCapabilities(hierarchy, itemId, isMatch, state: false);
-
-            return isTargetNode ? potentialTFM : null;
-
-            bool isMatch(ReadOnlyMemory<char> capability, bool _)
+            if (TryGetFlags(item, out var flags) &&
+                flags.Contains("TargetNode"))
             {
-                if (ReadOnlyMemoryOfCharComparer.Equals("TargetNode".AsSpan(), capability))
-                {
-                    isTargetNode = true;
-                }
-                else if (capability.Span.StartsWith("$TFM:".AsSpan()))
-                {
-                    potentialTFM = capability["$TFM:".Length..].ToString();
-                }
+                const string prefix = "$TFM:";
+                var flag = flags.FirstOrDefault(f => f.StartsWith(prefix));
 
-                return false;
+                return flag?.Substring(prefix.Length);
             }
+
+            return null;
         }
 
         private static bool NestedHierarchyHasProjectTreeCapability(IVsHierarchyItem item, string capability)
         {
-            var hierarchy = item.HierarchyIdentity.NestedHierarchy;
-            var itemId = item.HierarchyIdentity.NestedItemID;
+            if (TryGetFlags(item, out var flags))
+                return flags.Contains(capability);
 
-            return MatchProjectTreeCapabilities(hierarchy, itemId, isMatch, capability.AsMemory());
-
-            static bool isMatch(ReadOnlyMemory<char> capability, ReadOnlyMemory<char> capabilityToMatch)
-            {
-                return ReadOnlyMemoryOfCharComparer.Equals(capabilityToMatch.Span, capability);
-            }
+            return false;
         }
 
-        private static bool MatchProjectTreeCapabilities<T>(IVsHierarchy hierarchy, uint itemId, Func<ReadOnlyMemory<char>, T, bool> isMatch, T state)
+        public static bool TryGetFlags(IVsHierarchyItem item, out ProjectTreeFlags flags)
         {
-            if (hierarchy.GetProperty(itemId, (int)__VSHPROPID7.VSHPROPID_ProjectTreeCapabilities, out var capabilitiesObj) == VSConstants.S_OK)
+            if (item.HierarchyIdentity.IsRoot)
             {
-                var capabilitiesString = (string)capabilitiesObj;
+                if (item.HierarchyIdentity.NestedHierarchy is IVsBrowseObjectContext { UnconfiguredProject.Services.ProjectTreeService.CurrentTree.Tree: { } tree })
+                {
+                    flags = tree.Flags;
+                    return true;
+                }
+            }
+            else
+            {
```

Comment:
Do we need this block? Looks to me that if the item is the root, the `while` loop below will behave the same as the `if`. I think they can be merged into one block, and the code de-duplicated.

Reply:
If I'm wrong, then the comment should still be hoisted up to above the `if`, as it applies to both blocks.

Reply:
If I didn't do this, then if the passed in item was a root, the TryFind call would end up failing.

I don't object to moving the comment, but it seems most applicable here as this is the code that walks up the tree to find the appropriate object on which to create the IVsBrowseObjectContext

Reply:
Ok I see. The root will never be part of the dependencies tree, so for this usage it's fine to return no flags for the root. But no issue to leave as-is. If you want to keep it, maybe add a comment that `TryFind(IntPtr)` doesn't work for the root node, as it's unclear why it's given special treatment.